storage/innobase/buf/buf0buf.cc

```diff
@@ -742,6 +742,214 @@ bool buf_page_is_corrupted(bool check_lsn, const byte *read_buf,
+                  "insufficient fds");
+    std::string memcgroup{"/sys/fs/cgroup"};
+    std::string cgroup;
+    std::ifstream selfcgroup("/proc/self/cgroup");
+    std::getline(selfcgroup, cgroup, '\n');
+
+    cgroup.erase(0, 3); // Remove "0::"
+    memcgroup+= cgroup + "/memory.pressure";
+
+    m_num_fds= 0;
+    for (auto trig= std::begin(m_triggers); trig!= std::end(m_triggers); ++trig)
+    {
+      if ((m_fds[m_num_fds].fd=
+             open(memcgroup.c_str(), O_RDWR | O_NONBLOCK | O_CLOEXEC)) < 0)
+      {
+        switch (errno) {
+        case EPERM:
+        /* https://lore.kernel.org/all/CAMw=ZnQ56cm4Txgy5EhGYvR+Jt4s-KVgoA9_65HKWVMOXp7a9A@mail.gmail.com/T/#m3bd2a73c5ee49965cb73a830b1ccaa37ccf4e427 */
+          sql_print_information("InnoDB: Failed to initialize memory pressure EPERM, "
+                                "file permissions, CAP_SYS_RESOURCE, or Kernel fix required"
+				"(applies to 6.5, 6.6 kernels)");
+          break;
+        default:
+          sql_print_information("InnoDB: Failed to initialize memory pressure: %s",
+                                strerror(errno));
+        }
+        shutdown();
+        return false;
+      }
+      my_register_filename(m_pipe[0], memcgroup.c_str(), FILE_BY_OPEN, 0, MYF(0));
+      ssize_t slen= strlen(*trig);
+      if (write(m_fds[m_num_fds].fd, *trig, slen) < slen)
+      {
+        sql_print_warning("InnoDB: Failed create trigger for memory pressure \"%s\"", *trig);
+        my_close(m_fds[m_num_fds].fd, MYF(MY_WME));
+        continue;
+      }
+      m_fds[m_num_fds].events= POLLPRI;
+      m_num_fds++;
+    }
+    if (m_num_fds < 1)
+      return false;
+
+    if (pipe2(m_pipe, O_CLOEXEC | O_DIRECT))
+    {
+      sql_print_warning("InnoDB: No memory pressure - can't create pipe");
+      shutdown();
+      return false;
+    }
+    my_register_filename(m_pipe[0], "mem_pressure_read", FILE_BY_DUP, 0, MYF(0));
+    my_register_filename(m_pipe[1], "mem_pressure_write", FILE_BY_DUP, 0, MYF(0));
+    m_fds[m_num_fds].fd= m_pipe[0];
+    m_fds[m_num_fds].events= POLLIN;
+    m_num_fds++;
+    m_thd= std::thread(pressure_routine, this);
+    return true;
+  }
+
+  void shutdown()
+  {
+    while (m_num_fds)
+    {
+      m_num_fds--;
+      my_close(m_fds[m_num_fds].fd, MYF(MY_WME));
+    }
+    if (m_pipe[1] >= 0)
+      my_close(m_pipe[1], MYF(MY_WME));
+  }
+
+  static void pressure_routine(mem_pressure *m)
+  {
+    if (my_thread_init())
+      return;
+
+    ulonglong last= microsecond_interval_timer() - max_interval_us;
+    while (1)
+    {
+      if (poll(&m->m_fds[0], m->m_num_fds, -1) < 0)
+      {
+        if (errno == EINTR)
+          continue;
+        else
+        {
+          sql_print_warning("InnoDB: memory pressure poll error %s",
+                            strerror(errno));
+          m->shutdown();
+          break;
+        }
+      }
+      for (pollfd &p : st_::span<pollfd>(m->m_fds, m->m_num_fds))
+      {
+        if (p.revents & POLLPRI)
```

Comment:
Do we really need the loop here? Only for this first pipe we subscribed to `POLLPRI` events, and only for the second pipe we subscribed to `POLLIN`. Could we just check
```c++
if (m->fds[0].revents & POLLPRI) { … }
if (m->fds[1].revents & POLLIN) { … }
```

Reply:
So the first 2 are POLLPRI - two triggers, and I've written the `setup` to tolerate one of the failing to initialize (on EPERM)., and then its the last that's POLLIN only. So yes, we've over checking slightly, but on 3 memory structures that are already in cache that aren't hotly used I was going to readability.

Reply:
OK, then let’s retain the loop.